### PR TITLE
[RISCV][GISEL] Add support for scalable vector types in lowerReturnVal

### PR DESCRIPTION
**Pull request by michaelmaitland**
Scalable vector types from LLVM IR are lowered into physical vector registers in MIR based on calling convention for return instructions.

### CODE REVIEW

File: llvm/lib/Target/RISCV/GISel/RISCVCallLowering.cpp [hunk 1]
```diff
@@ -301,6 +301,27 @@ struct RISCVCallReturnHandler : public RISCVIncomingValueHandler {
 RISCVCallLowering::RISCVCallLowering(const RISCVTargetLowering &TLI)
     : CallLowering(&TLI) {}
 
+/// Return true if scalable vector with ScalarTy is legal for lowering.
+static bool isLegalElementTypeForRVV(Type *EltTy,
```
Comment (arsenm) [inline llvm/lib/Target/RISCV/GISel/RISCVCallLowering.cpp:305]: Why do you need to filter the IR like this? I would hope anything unhandled would naturally fail and hit the fallback

File: llvm/test/CodeGen/RISCV/GlobalISel/irtranslator/vec-ret.ll [hunk 2]
```diff
@@ -0,0 +1,809 @@
...
+  ; RV32-NEXT:   $v8m2 = COPY [[DEF]](<vscale x 2 x s64>)
+  ; RV32-NEXT:   PseudoRET implicit $v8m2
+  ;
+  ; RV64-LABEL: name: test_ret_nxv2f64
+  ; RV64: bb.1.entry:
+  ; RV64-NEXT:   [[DEF:%[0-9]+]]:_(<vscale x 2 x s64>) = G_IMPLICIT_DEF
+  ; RV64-NEXT:   $v8m2 = COPY [[DEF]](<vscale x 2 x s64>)
+  ; RV64-NEXT:   PseudoRET implicit $v8m2
+entry:
+  ret <vscale x 2 x double> undef
```
Comment (arsenm) [inline llvm/test/CodeGen/RISCV/GlobalISel/irtranslator/vec-ret.ll:584]: Non-undef return values are probably of more value to test with. A load is probably the best source

File: llvm/lib/Target/RISCV/GISel/RISCVCallLowering.cpp (hunk 1 above)
Comment (topperc) [reply llvm/lib/Target/RISCV/GISel/RISCVCallLowering.cpp:305]: The other types aren't listed as legal in SelectionDAG so getRegisterTypeForCallingConv() returns the scalar type or something and I think bad things happened from there. SelectionDAG doesn't fail gracefully either.

File: llvm/test/CodeGen/RISCV/GlobalISel/irtranslator/vec-ret.ll (hunk 2 above)
Comment (michaelmaitland) [reply llvm/test/CodeGen/RISCV/GlobalISel/irtranslator/vec-ret.ll:584]: At this point in time, we have not done any work to support scalable vector loads or stores in the IR translator. That is why I went with undef.

File: llvm/lib/Target/RISCV/GISel/RISCVCallLowering.cpp [hunk 3]
```diff
@@ -336,6 +357,10 @@ static bool isSupportedReturnType(Type *T, const RISCVSubtarget &Subtarget) {
     return true;
   }
 
+  if (T->isVectorTy() && Subtarget.hasVInstructions() && T->isScalableTy() &&
```
Comment (topperc) [inline llvm/lib/Target/RISCV/GISel/RISCVCallLowering.cpp:360]: Isn't `isSupportedReturnType` used by lowerCall too?

Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/GISel/RISCVCallLowering.cpp:360]: I think that’s a different function.

Comment (topperc) [reply llvm/lib/Target/RISCV/GISel/RISCVCallLowering.cpp:360]: What's a different function?

isSupportedReturnType has 2 callers just like isSupportedArgumentType. You're only testing one of the callers.

Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/GISel/RISCVCallLowering.cpp:360]: My bad. Updated.

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**